File: engine/lib/phx/script/meta/EventType.lua
---@meta

-- List of the event types used in the engine.
-- In Lua scripts should be used as an event id.
-- To extend it in Lua scripts call `EventType.AddEventTypes({"MyEventType1", "MyEventType2"})` function.
---@enum EventType
EventType = {
    -- Before physics update event
    PreSim = 0,
    -- Physics update event
    Sim = 1,
    -- After physics update event
    PostSim = 2,
    -- Before frame render event
    PreRender = 3,
    -- Frame render event
    Render = 4,
    -- After frame render event
    PostRender = 5,
    -- Before input handling event
    PreInput = 6,
    -- Input handling event
    Input = 7,
    -- After input handling event
    PostInput = 8,
    -- Specifies number of engine event types
    EngineEventTypesCount = 9,
}

File: engine/lib/phx/src/event_bus/event_type.rs
use super::EventId;

/// List of the event types used in the engine.
/// In Lua scripts should be used as an event id.
/// To extend it in Lua scripts call `EventType.AddEventTypes({"MyEventType1", "MyEventType2"})` function.
// NOTE: Use same type in 'repr' as in EventId.
#[luajit_ffi_gen::luajit_ffi(repr = "u16")]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EventType {
    // Before physics update event
    PreSim,
    // Physics update event
    Sim,
    // After physics update event
    PostSim,
    // Before frame render event
    PreRender,
    // Frame render event
    Render,
    // After frame render event
    PostRender,
    // Before input handling event
    PreInput,
    // Input handling event
    Input,
    // After input handling event
    PostInput,

    /// Specifies number of engine event types
    EngineEventTypesCount, // !!! SHOULD BE THE LAST ENUM VARIANT !!!
}

impl EventType {
    pub fn index(&self) -> EventId {
        *self as EventId
    }
}
